src/NBitcoin/ConsensusChainIndexer.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace NBitcoin
+{
+    public class ConsensusChainIndexer
+    {
+        private readonly object lockObject = new object();
+        private readonly Dictionary<int, ChainedHeader> blocksByHeight = new Dictionary<int, ChainedHeader>();
+        private readonly Dictionary<uint256, ChainedHeader> blocksById = new Dictionary<uint256, ChainedHeader>();
+
+        public Network Network { get; }
+        public ChainedHeader Tip { get; private set; }
+
+        public int Height => this.Tip.Height;
+        public ChainedHeader Genesis => this.GetBlock(0);
+
+        public ConsensusChainIndexer(Network network)
+        {
+            this.Network = network;
+
+            this.Initialize(new ChainedHeader(network.GetGenesis().Header, network.GetGenesis().GetHash(), 0));
+        }
+
+        public ConsensusChainIndexer(Network network, ChainedHeader chainedHeader) 
+        {
+            this.Network = network;
+
+            this.Initialize(chainedHeader);
+        }
+
+        public void Initialize(ChainedHeader chainedHeader)
+        {
+            this.blocksById.Clear();
+
+            lock (this.lockObject)
+            {
+                ChainedHeader iterator = chainedHeader;
+
+                while (iterator != null)
+                {
+                    this.blocksById.Add(chainedHeader.HashBlock, chainedHeader);
+                    this.blocksByHeight.Add(chainedHeader.Height, chainedHeader);
+
+                    if (chainedHeader.Height == 0)
+                    {
+                        if (chainedHeader.HashBlock != iterator.HashBlock)
+                            throw new InvalidOperationException("Wrong network");
+                    }
+
+                    iterator = iterator.Previous;
+                }
+
+                this.Tip = chainedHeader;
+            }
+        }
+
+
+        /// <summary>
+        /// Returns the first chained block header that exists in the chain from the list of block hashes.
+        /// </summary>
+        /// <param name="hashes">Hash to search for.</param>
+        /// <returns>First found chained block header or <c>null</c> if not found.</returns>
+        public ChainedHeader FindFork(IEnumerable<uint256> hashes)
+        {
+            if (hashes == null)
+                throw new ArgumentNullException("hashes");
+
+            // Find the first block the caller has in the main chain.
+            foreach (uint256 hash in hashes)
+            {
+                ChainedHeader mi = this.GetBlock(hash);
```

Comment:
Variable name could be improved, what is `mi`?

Reply:
ha yes will fix, it was a copy paste.